### PR TITLE
feat: add advanced RocksDB interface (Iterator, DeleteRange, WriteBatch, Snapshots)

### PR DESCRIPTION
**Pull request by SoumojitDalui**
## Dependencies

This PR depends on #60 (core RocksDB interface PR). Please merge that first.

## Summary

Adds four advanced RocksDB-compatible features on top of the core interface:

### Iterator (BufferedIterator)
- `IIterator` interface with Seek, SeekToFirst, SeekToLast, Next, Prev, Valid, key, value
- Buffer-based implementation: scan results loaded into memory on Seek, cursor moves through buffer
- 12 tests covering forward/reverse iteration, boundary conditions, isolation, large datasets

### DeleteRange
- `ITable::DeleteRange(txn, start_key, end_key, deleted_count)` 
- Scan-then-delete strategy, atomic within the enclosing OCC transaction
- 15 tests covering basic ranges, atomicity, edge cases, performance, concurrency

### WriteBatch
- `IWriteBatch` interface with Put, Delete, DeleteRange, Count, Clear, Write
- Queue-then-apply strategy: operations accumulated in memory, applied atomically via single OCC transaction
- Pre-encodes all values to avoid StringWrapper pointer aliasing bug
- Also fixes DB::InitThread() for worker threads in single-node mode
- 13 tests covering basic ops, atomicity, cross-table batches, performance, concurrency

### Snapshots (EagerSnapshot)
- `ISnapshot` interface with Get, Exists, NewIterator, GetSequenceNumber
- Per-table lazy buffering: O(1) creation, O(N) first-access per table
- True point-in-time consistency within a single table
- Limitation: no cross-table consistency (documented in docs/snapshot_design.md)
- 11 tests covering basic reads, isolation from concurrent writes, lifecycle

## Known Limitations

- Iterator: buffer-based (O(N) memory on Seek), no SeekForPrev
- DeleteRange: O(N) scan+delete, not O(1) tombstone like RocksDB
- WriteBatch: large batches (10K+) slower due to OCC validation overhead; sweet spot is 100-1000 ops
- Snapshots: per-table consistency only, not cross-table; O(N) memory per buffered table

## Test Results

- All existing CI suites pass (no regressions)
- 12 iterator tests pass
- 15 DeleteRange tests pass
- 13 WriteBatch tests pass
- 11 snapshot tests pass

### CODE REVIEW

File: src/mako/snapshot.hh [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+#include <map>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace mako {
+
+/**
+ * SnapshotIterator - Iterator over a buffered snapshot table.
```
Comment (shenweihai1) [inline src/mako/snapshot.hh:40]: Snapshot has to be implemented via global vector watermark cut (slow down the reclaim of versions); and coordinate across shards

File: src/mako/write_batch.hh [hunk 2]
```diff
@@ -0,0 +1,152 @@
+#pragma once
```
Comment (shenweihai1) [inline src/mako/write_batch.hh:1]: WriteBatch op is better to let user to do it in application layers, to avoid nested transactions
